header-rewrite-filter/header_processor.cc

```diff
@@ -53,6 +53,46 @@ namespace HeaderRewriteFilter {
         return absl::OkStatus();
     }
 
+    absl::Status AppendHeaderProcessor::parseOperation(std::vector<absl::string_view>& operation_expression, std::vector<absl::string_view>::iterator start) {
+        if (operation_expression.size() < Utility::APPEND_HEADER_MIN_NUM_ARGUMENTS) {
+            return absl::InvalidArgumentError("not enough arguments for set-header");
+        }
+
+        // TODO: remove
+        start++;
+
+        // parse key and call setKey
+        try {
+            const absl::string_view key = operation_expression.at(2); // TODO: use start
```

Comment:
Nit: would be good to unify the behaviour to use iterator `operator+` vs `.at`, to match `SetHeaderProcessor::parseOperation`, for consistency

Reply:
I also notice, if this is changed to only use the iterator and not `operation_expression.at`, you don't need the `operation_expression` at all. It's only ever used for the `end` iterator, plus the argument check with `.size()`.

For both this and `SetHeaderProcessor::parseOperation`, you could change the signature to be
```
absl::Status AppendHeaderProcessor::parseOperation(std::vector<absl::string_view>::iterator start, std::vector<absl::string_view>::iterator end) {
```

Which makes the code a little clearer and safer. It makes it clear to users that they should only ever operate within the bounds of `(start, end)`, and not do random other things with the vector. And it makes modifying the vector in certain ways impossible.

You can do the arg check with `std::distance(start, end)`

Even better: since this doesn't modify the vector, and I don't think it ever makes sense for parsing to modify it, you can change it to `cbegin()`/`cend()` and `const_iterator`, so parseOperation _can't_ modify it.

Reply:
I really like this suggestion!

On the topic of `string_view`, I just wanted to double check that I'm using `string_view`, `std::string`, and `iterator` in the right way. Correct me if I'm wrong: it's okay to pass a `string_view` iterator from the HTTP filter to a `Processor` and assume that the contents of the `string_view` are not out of scope, but any part of the `string_view` that I want to store in the `Processor` object needs to get copied?

If that's the case, then I'm guessing it makes sense to pass `const_iterator` for start and end but otherwise will I have to pass a copy of the `string_view` or somehow create a `std::unique_ptr` and `std::move` the `string_view` into the scope of the `Processor`?

Reply:
> pass a string_view iterator

These are vector iterators, not a string_view iterators. 

> Correct me if I'm wrong: it's okay to pass a string_view iterator from the HTTP filter to a Processor and assume that the contents of the string_view are not out of scope, but any part of the string_view that I want to store in the Processor object needs to get copied?

That's all correct. But you don't have `string_view::iterator`s here. You could, if you want, but that wasn't what I was suggesting.

> create a std::unique_ptr and std::move the string_view

I don't think that would make sense, because `string_view` doesn't own its contents. It's a reference to memory owned by something else (Usually the `string` or `char*` that the view was constructed from). `unique_ptr` and `std::move` are for managing memory that you own. The primary reason `unique_ptr` and `shared_ptr` exist is to call `delete` when all copies go out of scope; if you didn't need to call delete to free memory (and/or call destructors), you wouldn't need them.

Think of `std::string` as a `unique_ptr<char *>`. It does some other things, like dynamically allocating more memory when you append to the string. But that's basically what it is: managing memory of the char*.

And so, think of `string_view` as calling 

```
unique_ptr<char*> pseudo_string = std::unique_ptr<char*>(new char[42]);
const char* pseudo_string_view = pseudo_string.get()
```

You can always call `my_unique_ptr.get()` to get the raw underlying pointer. But you have to be careful when you do, that you never access it after the `my_unique_ptr` object goes out of scope. 

`std::string` and `std::string_view`/`absl::string_view` are almost identical to `unique_ptr` and getting the raw underlying pointer via `unique_ptr::get`. (with the exception that you don't have to call `new` to allocate the memory, std::string does that for you, and copies whatever `char*` you give it)

Reply:
But you have vector iterators here, not string/view iterators. So I think all your existing code is fine, and it should "just work" to replace `std::vector<absl::string_view>& operation_expression` and `operation_expression.size() and `operation_expression.end()` with `std::vector<absl::string_view>::iterator end` and `std::distance(start, end)`

Without changing any of the string/string_view code

Reply:
Will address in a separate PR focused on code cleanup.